art/attacks/evasion/projected_gradient_descent/projected_gradient_descent_numpy.py

```diff
@@ -259,43 +255,62 @@ def generate(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> n
             targets = self._set_targets(x, y)
 
             # Start to compute adversarial examples
-            adv_x_best = None
-            rate_best = None
-
-            for _ in trange(
-                max(1, self.num_random_init), desc="PGD - Random Initializations", disable=not self.verbose
-            ):
-                adv_x = x.astype(ART_NUMPY_DTYPE)
-
-                for i_max_iter in trange(self.max_iter, desc="PGD - Iterations", leave=False, disable=not self.verbose):
-                    adv_x = self._compute(
-                        adv_x,
-                        x,
-                        targets,
-                        mask,
-                        self.eps,
-                        self.eps_step,
-                        self._project,
-                        self.num_random_init > 0 and i_max_iter == 0,
-                    )
-
-                if self.num_random_init > 1:
-                    rate = 100 * compute_success(
-                        self.estimator, x, targets, adv_x, self.targeted, batch_size=self.batch_size,  # type: ignore
-                    )
-                    if rate_best is None or rate > rate_best or adv_x_best is None:
-                        rate_best = rate
-                        adv_x_best = adv_x
-                else:
-                    adv_x_best = adv_x
+            adv_x = x.astype(ART_NUMPY_DTYPE)
+
+            for batch_id in range(int(np.ceil(x.shape[0] / float(self.batch_size)))):
+                for rand_init_num in trange(
+                    max(1, self.num_random_init), desc="PGD - Random Initializations", disable=not self.verbose
+                ):
+                    batch_index_1, batch_index_2 = batch_id * self.batch_size, (batch_id + 1) * self.batch_size
+                    batch_index_2 = min(batch_index_2, x.shape[0])
+                    batch = x[batch_index_1:batch_index_2]
+                    batch_labels = targets[batch_index_1:batch_index_2]
+                    mask_batch = mask
```

Comment:
Can we combine this line with the code in line 270?

Reply:
Hi Beat and thanks for the comments! This is following the coding pattern used in fast_gradient.py so perhaps we can keep it as it is for consistency between code.